### ISSUE TITLE
{{url}} is bypassing the security checks in the download allow/deny function

### ISSUE THREAD

**Issue by tanis2000** [CollectionFS/Meteor-CollectionFS#269]
I have the following allow/deny rules for a collectionFS:

```
MessagesFS.allow({
  insert: function (userId, file) {
    return userId && file.owner === userId;
  },
  update: function (userId, file, fields, modifier) {
    return userId && file.owner === userId;
  },
  remove: function (userId, file) {
    return userId && file.owner === userId;
  },
  download: function (userId, file) {
    return userId && file.owner === userId;
  }
});
```

In my template I use the {{url}} helper to get an URL to the file and it works fine for the currently logged in user.

But if I copy that URL and paste it in another browser with no user signed in, I can still access the file. 

Is there a way to block this behavior?

**Comment by adamgins** [comment]
HI @tanis2000 @aldeed @raix how goes?
Just wondering did this actually get implemented... ie is there a way I can expire the token easily or a way to stop/limit folk coping and pasting the URL?

I do understand that in theory, this is really just a deterrent for folk as it's the person with access who may send a link to some info and there is really nothing stopping them downloading the doc and giving it to another unauthorized user.

**Comment by tanis2000** [comment]
Hey @adamgins, this got implemented in a way that lets you set an expire token. I useitto provide links that expire after one minute as they contain sensitive information but the data is actually being provided theough the s3 adapter and to anonymous users who got the link through email. The link I'm sending is set to transfer the user to a landing page where I can do a couple of checks and afterwards I give them the actual link to the file with the expiration time set.

If you need a sample I'll dig it up as soon as I get in front of a desktop. Cheers!

**Comment by adamgins** [comment]
@tanis2000 thanks. Yep a sample would be great. 

As I mentioned, I am not sure it solves the sensitive info issue... ie if someone downloads the file and forwards the link the info is still out there... also if they forward/tweet the link and it's clicked within a minute, if I am understanding correctly they'd get access, right? Or do the other checks prevent this? Just wondering if there's a way to share/force login on the link so only the user gets access to the s3 file?

**Comment by tanis2000** [comment]
@adamgins no, this solution doesn't cover security as the link will be good for everyone until it expires. 

What I'm adding here is the `expiration` property of the token. But you can instead set the `authToken` property to that same token and set it to the loginToken of the user that should have access to that file.

If you do it that way, CFS automatically checks that the user can actually access the file.
You can learn more by having a look at the code there: https://github.com/CollectionFS/Meteor-CollectionFS/blob/master/packages/access-point/access-point-server.js#L313

Adding a link with the expiration token is as easy as adding this to your template:

    <a href="{{url auth=60}}" class="btn btn-primary btn-mini">{{name}}</a>

If you set `auth` to a string and actually set it to your loginToken, you get what I was referring to above.

Again, you can learn more about it by having a look at the code there: https://github.com/CollectionFS/Meteor-CollectionFS/blob/master/packages/access-point/access-point-common.js#L115

**Comment by niranjans** [comment]
Hi @tanis2000 
I was trying to follow this thread and try to make sense of it but I'm new to using this package and I'm not sure how to go about making a timebound-url expiry work. 

What I'd like is to generate and share a file-url which expires in a few seconds (let's say 60). Till the time the url is valid, even an anonymous/non-logged in user should be able to access the file data. 

So this helper does just that - `{{url auth=60}}`
However, how do I check in the allow/deny download function to check if the token is valid and is not expired yet?

```
FileObjects.deny({
 download: function(userId, fileObj){
  // ----- What to check for?
  // userId can be null since even anonymous users can download this file while it is valid
  return false;
 }
});
```

Any pointers would help. 

Thanks
